daemon/network.go

```diff
@@ -99,15 +100,51 @@ func (daemon *Daemon) getAllNetworks() []libnetwork.Network {
 	return daemon.netController.Networks()
 }
 
-func isIngressNetwork(name string) bool {
-	return name == "ingress"
+type ingressJob struct {
+	create *clustertypes.NetworkCreateRequest
+	ip     net.IP
 }
 
-var ingressChan = make(chan struct{}, 1)
+var (
+	ingressWorkerOnce  sync.Once
+	ingressJobsChannel chan *ingressJob
+	ingressWorkerStop  chan struct{}
+	ingressID          string
+	ingress            sync.Mutex
```

Comment:
> SetupIngress / ReleaseIngress are using ingressWorkerOnce

Not sure I follow. `Once.Do(f)` is designed to not allow more than one execution of `f`, even in concurrent environment. I do not think there is any need to hold a lock during a call to Once.Do. 

Being called in a once do construct, `startIngressWorker` is guaranteed to be called once by one thread.

Still, as you said, `startIngressWorker` initializes the two channels, and `stopIngressNetwork`, which also accesses the channels, could be concurrently called. To protect against this, both `startIngressWorker` and `stopIngressNetwork` are now locking around the access.

Yes, then it spawns the goroutine which listen on those channels, with no protection.

The correct thing here is probably to define a `type ingressWorker struct` with `start`, `stop`, `enqueue` methods, so that it can control the access to and lifecycle of its channels.